File: services/api/schemas/outbox.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from services.api.outbox.status import (
    OutboxEventStatus,
)


class OutboxEventResponse(BaseModel):
    model_config = ConfigDict(
        use_enum_values=True,
    )

    event_id: UUID
    aggregate_type: str
    aggregate_id: str
    event_type: str
    schema_version: int
    source: str
    occurred_at: datetime
    created_at: datetime
    published_at: datetime | None
    attempt_count: int
    next_attempt_at: datetime
    last_error: str | None
    locked_by: str | None
    locked_until: datetime | None
    status: OutboxEventStatus


class OutboxEventListResponse(BaseModel):
    model_config = ConfigDict(
        use_enum_values=True,
    )

    items: list[OutboxEventResponse]
    count: int
    status_filter: OutboxEventStatus


class OutboxSummaryResponse(BaseModel):
    total_count: int
    pending_count: int
    published_count: int
    failed_count: int
    unpublished_count: int
    oldest_unpublished_at: datetime | None

File: frontend/web/src/types/outbox.ts
export type OutboxEventStatus =
  | "pending"
  | "published"
  | "failed";

export type OutboxStatusFilter =
  | "all"
  | OutboxEventStatus;

export interface OutboxEvent {
  event_id: string;
  aggregate_type: string;
  aggregate_id: string;
  event_type: string;
  schema_version: number;
  source: string;
  occurred_at: string;
  created_at: string;
  published_at: string | null;
  attempt_count: number;
  next_attempt_at: string;
  last_error: string | null;
  locked_by: string | null;
  locked_until: string | null;
  status: OutboxEventStatus;
}

export interface OutboxEventListResponse {
  items: OutboxEvent[];
  count: number;
  status_filter: OutboxStatusFilter;
}

export interface OutboxSummaryResponse {
  total_count: number;
  pending_count: number;
  published_count: number;
  failed_count: number;
  unpublished_count: number;
  oldest_unpublished_at: string | null;
}
